draft-ietf-tls-tls13.md

```diff
@@ -3390,14 +3424,29 @@ TLS-compliant application MUST implement the following TLS extensions:
 
   * Signature Algorithms ("signature_algorithms"; {{signature-algorithms}})
   * Negotiated Groups ("supported_groups"; {{negotiated-groups}})
+  * Client Key Share ("client_key_share"; {{client-key-share}})
+  * Pre-Shared Key Extension ("pre_shared_key"; {{pre-shared-key-extension}})
+  * Server Name Indication ("server_name"; Section 3 of {{RFC6066}})
+
+All implementations MUST send and use these extensions when offering
+applicable cipher suites:
+
+  * "signature_algorithms" is REQUIRED for certificate authenticated cipher suites
+  * "supported_groups" and "client_key_share" are REQUIRED for DHE or ECDHE cipher suites
+  * "pre_shared_key" is REQUIRED for PSK cipher suites
+
+When negotiating use of applicable cipher suites, endpoints MUST abort the
+connection with a "missing_extension" alert if the required extension was
+not provided. Any endpoint that receives any invalid combination of cipher
+suites and extensions MAY abort the connection with a "missing_extension"
+alert, regardless of negotiated parameters.
 
-All implementations MUST use the "signature_algorithms" extension when
-offering and negotiating certificate authenticated cipher suites.
-All implementations MUST use the "supported_groups" extension when
-offering and negotiating DHE or ECDHE cipher suites.
+Additionally, all implementations MUST support use of the "server_name"
+extension with applications capable of using it.
+Servers MAY require clients to send a valid "server_name" extension.
+Servers requiring this extension SHOULD respond to a ClientHello
+lacking a "server_name" extension with a fatal "missing_extension" alert.
```

Comment:
I don't think that this is right.  If we ever want to have SNI encryption, then this would be a problem for that.